network/alsp/manager/manager.go

```diff
@@ -237,6 +259,77 @@ func (m *MisbehaviorReportManager) HandleMisbehaviorReport(channel channels.Chan
 	lg.Debug().Msg("misbehavior report submitted")
 }
 
+// startHeartbeatTicks starts the heartbeat ticks ticker to tick at the given intervals. It is a blocking function, and
+// should be called in a separate goroutine. It returns when the context is canceled. Hearbeats are recurring events that
+// are used to perform periodic tasks.
+// Args:
+//
+//	ctx: the context.
+//	interval: the interval between two ticks.
+//
+// Returns:
+//
+//	none.
+func (m *MisbehaviorReportManager) startHeartbeatTicks(ctx irrecoverable.SignalerContext, interval time.Duration) {
+	ticker := time.NewTicker(interval)
+	m.logger.Info().Dur("interval", interval).Msg("starting heartbeat ticks")
+	defer ticker.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			m.logger.Debug().Msg("heartbeat ticks stopped")
+			return
+		case <-ticker.C:
+			m.logger.Trace().Msg("new heartbeat ticked")
+			m.onHeartbeat(ctx)
+		}
+	}
+}
+
+// onHeartbeat is called upon a startHeartbeatTicks. It encapsulates the recurring tasks that should be performed
+// during a heartbeat, which currently includes decay of the spam records.
+// Args:
+//
+//	ctx: the context.
+//
+// Returns:
+//
+//	none.
+func (m *MisbehaviorReportManager) onHeartbeat(ctx irrecoverable.SignalerContext) {
+	allIds := m.cache.Identities()
+
+	for _, id := range allIds {
+		penalty, err := m.cache.Adjust(id, func(record model.ProtocolSpamRecord) (model.ProtocolSpamRecord, error) {
+			if record.Penalty > 0 {
+				// sanity check; this should never happen.
+				return record, fmt.Errorf("illegal state: spam record %x has positive penalty %f", id, record.Penalty)
+			}
+			if record.Decay <= 0 {
+				// sanity check; this should never happen.
+				return record, fmt.Errorf("illegal state: spam record %x has non-positive decay %f", id, record.Decay)
+			}
+
+			// each time we decay the penalty by the decay speed, the penalty is a negative number, and the decay speed
+			// is a positive number. So the penalty is getting closer to zero.
+			// We use math.Min() to make sure the penalty is never positive.
+			record.Penalty = math.Min(record.Penalty+record.Decay, 0)
+			return record, nil
+		})
+
+		// any error here is fatal because it indicates a bug in the cache. All ids being iterated over are in the cache,
+		// and adjust function above should not return an error unless there is a bug.
+		if err != nil {
+			ctx.Throw(fmt.Errorf("failed to decay spam record %x: %w", id, err))
```

Comment:
it's a bit pedantic, but I think `onHeartbeat` should return an error, and the throw should happen in the `startHeartbeatTicks` loop

Reply:
https://github.com/onflow/flow-go/pull/4367/commits/b46a2b4a0d0fdab50a3da46208085ae8ead4842b